guides/policy/azure-authenticator/parseAzureAuthenticator.js

```diff
@@ -20,52 +20,67 @@
+    function getPolicyFromToken(encodedToken) {
         try {
             let decodedToken = jwt_decode(encodedToken);
-            let token = parseAzureToken(decodedToken);
+            let azureIdentity = parseAzureToken(decodedToken);
+            let identityAnnotation = azureIdentity.isUserIdentity ? "user-assigned-identity" : "system-assigned-identity";
 
             return `
             - !host
               id: ${hostId}
               annotations:
-                authn-azure/subscription-id: ${token.subscriptionId}
-                authn-azure/resource-group: ${token.resourceGroup}
+                authn-azure/subscription-id: ${azureIdentity.subscriptionId}
+                authn-azure/resource-group: ${azureIdentity.resourceGroup}
               # The following is optional
-                authn-azure/${token.isUserIdentity ? "user" : "system"}-assigned-identity: ${token.identity}
+                authn-azure/${identityAnnotation}: ${azureIdentity.identity}
             `;
         } catch (e) {
             console.error(e);
-            return "Failed to parse Azure AD Token"
+            return "Failed to parse Azure managed identity Token: \n" + e;
         }
     }
 
     function parseAzureToken(token) {
-        return token.xms_mirid.split('/').reduce((azureIdentity, cur, i, arr) => {
-            switch (cur) {
-                case "subscriptions":
-                    azureIdentity.subscriptionId = arr[i + 1];
-                    break;
-                case "resourcegroups":
-                    azureIdentity.resourceGroup = arr[i + 1];
-                    break;
-                case "providers":
-                    azureIdentity.isUserIdentity = arr.includes("Microsoft.ManagedIdentity", i + 1);
-                    azureIdentity.identity = azureIdentity.isUserIdentity ? arr[arr.length - 1] : token.oid;
-                    break;
-            }
-            return azureIdentity;
-        }, {});
+        let subscriptionRegex = /\bsubscriptions\/([^\/]+)/;
+        let subscription = findInXmsMirid("subscriptions", subscriptionRegex, token);
+
+        let resourceGroupsRegex = /\bresourcegroups\/([^\/]+)/;
+        let resourceGroup = findInXmsMirid("resourcegroups", resourceGroupsRegex, token);
+
+        let azureIdentity = {
+            subscriptionId: subscription,
+            resourceGroup: resourceGroup
+        };
+
+        let providersRegex = /\bproviders\/(.+)$/;
+        let providers = findInXmsMirid("providers", providersRegex, token);
+
+        let userIdentityRegex = /Microsoft\.ManagedIdentity\/(?:[^\/]+\/)*([^\/]+)/;
+        let userIdentity = userIdentityRegex.exec(providers);
+
+        azureIdentity.isUserIdentity = userIdentity && userIdentity.length > 1;
+        azureIdentity.identity = azureIdentity.isUserIdentity ? userIdentity[1] : token.oid;
+
+        return azureIdentity;
+    }
+
+    function findInXmsMirid(name, subscriptionRegex, token) {
```

Comment:
all "subscriptionRegex" values share the same template... why not give this parameter a better naming? and eliminate the need for "name" parameter?
maybe xmsMiridKey ?
Then you can construct the regex in the function